presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/WindowFilterPushDown.java

```diff
@@ -60,220 +64,149 @@ public PlanNode optimize(PlanNode plan, Session session, Map<Symbol, Type> types
-                Symbol rowNumberSymbol = Iterables.getOnlyElement(node.getWindowFunctions().entrySet()).getKey();
-                return WindowLimitExtractor.extract(filter.getFilterExpression().get(), rowNumberSymbol);
-            }
-            return Optional.empty();
-        }
-
-        private static boolean canOptimizeWindowFunction(WindowNode node)
-        {
-            if (node.getWindowFunctions().size() != 1) {
-                return false;
-            }
-            Symbol rowNumberSymbol = getOnlyElement(node.getWindowFunctions().entrySet()).getKey();
-            return isRowNumberSignature(node.getSignatures().get(rowNumberSymbol));
-        }
-
-        private static boolean filterContainsWindowFunctions(WindowNode node, Expression filterPredicate)
-        {
-            Set<Symbol> windowFunctionSymbols = node.getWindowFunctions().keySet();
-            Sets.SetView<Symbol> commonSymbols = Sets.intersection(DependencyExtractor.extractUnique(filterPredicate), windowFunctionSymbols);
-            return !commonSymbols.isEmpty();
+            return context.replaceChildren(node, ImmutableList.of(rewrittenSource));
         }
 
         @Override
-        public PlanNode visitLimit(LimitNode node, RewriteContext<Constraint> context)
+        public PlanNode visitLimit(LimitNode node, RewriteContext<Void> context)
         {
             // Operators can handle MAX_VALUE rows per page, so do not optimize if count is greater than this value
             if (node.getCount() >= Integer.MAX_VALUE) {
                 return context.defaultRewrite(node);
             }
-            Constraint constraint = new Constraint(Optional.of((int) node.getCount()), Optional.empty());
-            PlanNode rewrittenSource = context.rewrite(node.getSource(), constraint);
 
-            if (rewrittenSource != node.getSource()) {
-                return new LimitNode(idAllocator.getNextId(), rewrittenSource, node.getCount());
+            PlanNode rewrittenSource = context.rewrite(node.getSource());
+            if (rewrittenSource instanceof RowNumberNode) {
+                RowNumberNode rowNumberNode = (RowNumberNode) rewrittenSource;
+                if (rowNumberNode.getPartitionBy().isEmpty()) {
+                    return (rowNumberNode).withLimit(Ints.checkedCast(node.getCount()));
+                }
             }
 
-            return context.defaultRewrite(node);
+            if (rewrittenSource instanceof WindowNode) {
+                WindowNode windowNode = (WindowNode) rewrittenSource;
+                if (canOptimizeWindowFunction(windowNode) && windowNode.getPartitionBy().isEmpty()) {
+                    // verify that unordered row_number window functions are replaced by RowNumberNode
+                    verify(!windowNode.getOrderBy().isEmpty());
+                    return convertToTopNRowNumber(windowNode, Ints.checkedCast(node.getCount()));
```

Comment:
you can convert to TopNRowNumber even if partitionBy is not empty by just enforcing the Limit right afterwards. This should still help save a good amount of memory.